File: joao/131_palindrome_partitioning.py
class Solution(object):
    def partition(self, s):
        def is_palindrome(sub):
            return sub == sub[::-1]

        def backtrack(start, acc):
            if start == len(s):
                result.append(acc)
                return

            for end in range(start + 1, len(s) + 1):
                candidate = s[start:end]
                if is_palindrome(candidate):
                    backtrack(end, acc + [candidate])

        result = []
        backtrack(0, [])
        return result

File: julianoATS/131_palindromePartitioning.ts
function isPalindrome(s: string, left: number, right: number): boolean {
    while (left < right) {
        if (s[left] !== s[right]) return false;
        left++;
        right--;
    }

    return true;
};

function palindromePartitioning(result: string[][], start: number, path: string[], s: string): void {
    if (start === s.length) {
        result.push([...path]);
        return;
    }

    for (let end = start; end < s.length; end++) {
        if(!isPalindrome(s, start, end)) continue;

        path.push(s.substring(start, end + 1));
        palindromePartitioning(result, end + 1, path, s);
        path.pop();
    }
};

function partition(s: string): string[][] {
    const result: string[][] = [];

    palindromePartitioning(result, 0, [], s);

    return result;
};
